### PR TITLE
Core typing: compat/config/fields/main/packet

### PR DESCRIPTION
**Pull request by gpotter2**
Kick off Hinty (https://github.com/secdev/scapy/issues/2158):
- turn on Mypy strict mode (see [this](https://github.com/secdev/scapy/blob/2a914cf2e29e98adea88ed447e524a9970b2e451/.config/mypy/mypy_check.py#L39))
- add typings to 
  - packet.py
  - fields.py
  - config.py
  - main.py (strict)
  - \_\_init\_\_.py (strict)

**Implementations notes:**
- this tries to be as clean and extensible as possible: we use `Generic` classes to make clean `Field`s. This means any other field that inherits from `Field` will be able to mark what `Internal` and `Machine` types it uses.
- this PR fixes a few type errors that mypy triggered. For instance a function that would crash in `SndRcvList` and was moved to a parent class.
- I refactored `TimeSignedField` because it was easier than cleaning it up.
- This is fully compatible except one single thing: the use of `Generic` makes it impossible to use the keyword `cls=` as a key parameter. This keyword was only used in `PacketListField` (in other places, it was only used as an argument which is fine). Therefore I had to rename it to `pkt_cls`, but it's use was not that widespread because you would typically use it as an argument instead of a key argument.

To reproduce the issue with `Generic`:
```python
from typing import Generic, TypeVar
I = TypeVar("I")
class A(Generic[I]):
    def __init__(self, cls=1):
        pass

A(cls=1)  # Error
```

- **This is a big PR, but had to be done in one go because those files just keep referring to each other.** Making them one by one would mean adding a gazillion temporary `# type: ignore`. I hesitated to also do `utils.py`, but it's honnestly to big on its own and will need a separate PR.
- python 2 compatibility was added, but it's quite a nightmare. See the code. I sometimes wish we had drop it. In any case, when will want to drop it (likely in a few years), we'll have the great [`com2ann`](https://github.com/ilevkivskyi/com2ann) module to do it for us.

### CODE REVIEW

File: scapy/packet.py
Comment (gpotter2) [inline scapy/packet.py:1275]: I didn't understand the piece of code before this line (the for). Why do we use this for already? Can you get a field's value using `getlayer` ?!

File: scapy/compat.py
Comment (gpotter2) [reply scapy/compat.py:166]: The issue is that `compat.py` will cause import loops all over the place if we start importing other classes. How would you do that?

File: scapy/compat.py
Comment (gpotter2) [reply scapy/compat.py:218]: There's actually already `AnyStr` which does that. good idea

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:256]: No you can't for `__eq__`, `__new__`, `__add__`.... or anything that extends the object class. It makes sense, you can always try to use `==` on another type, so it should remain `Any`. I'll mark as resolved similar cases.

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:1179]: *int

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:1209]: *int

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:1799]: No. Any could be any field's internal value.

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:1839]: Don't think so. Not without adding an extra `Generic` in `FieldListField`. Should we do that?

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:2542]: No, as said above.

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:2713]: For the `pkt` yes. For the `x` however, I remember trying quite some time but eventually abandonning: it's a nightmare.

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:366]: What is the argument even used for?

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:565]: No, same reason than with `__eq__`

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:578]: Same

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:586]: same

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:1172]: Same

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:1183]: Same

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:1810]: Same

File: scapy/utils.py
Comment (gpotter2) [reply scapy/utils.py:196]: Well it can be a tuple, a list, a volatile value, a string or an int, (or anything really, it just falls back on `str()`).. :p

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:260]: same

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:206]: I tried but couldn't make this work. It will work for this function, but then in functions that extend `do_copy` it creates strange errors.

File: scapy/compat.py
Comment (guedou) [reply scapy/compat.py:166]: Will moving `raw()` to `scapy/packet.py` allow a better typing?

File: scapy/fields.py
Comment (guedou) [reply scapy/fields.py:256]: I am sad =/

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:1113]: Or literally any other value? The code calling it is, in summary:
```
for field in fields:
    fld, val = self.getfield_and_val(field)
    if is_valid_gen_tuple(val):
        [....]
```
 so it could really be anything :/

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:827]: What are those ?
`pyx.canvas.canvas` and `pyx.color.color` ? Can't find any refs for those :/

File: scapy/packet.py
Comment (gpotter2) [reply scapy/packet.py:790]: `pyx.path.pathitem` ? or `pyx.path.path` ?

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:1835]: oh you meant the first any

**Comment by guedou** [comment]
It seems that there is a conflict with `packet.py`. Could you fix before I review your last changes?

File: scapy/fields.py
Comment (guedou) [reply scapy/fields.py:256]: Is it a limitation of the Python typing system? With other typed languages, we will be able to restrict the type of `other`. That sounds strange.

File: scapy/compat.py
Comment (guedou) [inline scapy/compat.py:190]: Could we replace Any with AnyStr ?

File: scapy/fields.py
Comment (guedou) [inline scapy/fields.py:374]: Any should be replaced by Field

File: scapy/fields.py
Comment (guedou) [inline scapy/fields.py:400]: Any should be replaced by Field.

File: scapy/fields.py
Comment (guedou) [inline scapy/fields.py:433]: Any could be replaced by Field.

File: scapy/fields.py
Comment (guedou) [inline scapy/fields.py:554]: Does this work with Python2?

File: scapy/fields.py
Comment (guedou) [inline scapy/fields.py:664]: The first Any could be replaced by Union(AnyStr, List[AnyStr]).

File: scapy/fields.py
Comment (guedou) [inline scapy/fields.py:905]: It seems that Any can indeed by typed as int.

File: scapy/fields.py
Comment (guedou) [inline scapy/fields.py:1359]: Was this bug catched by mypy?

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:554]: Yes thanks to the `__getitem__` trick in `base_classes.py`

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:1359]: yes

Comment (guedou) [reply scapy/fields.py:1359]: Awesome!

File: scapy/fields.py
Comment (guedou) [reply scapy/fields.py:554]: Cool.

File: scapy/fields.py
Comment (guedou) [inline scapy/fields.py:2470]: Could Any be replaced by str?

File: scapy/compat.py
Comment (gpotter2) [reply scapy/compat.py]: I've found a way. see https://github.com/secdev/scapy/pull/2671/commits/3ffe840dc6ed80678eec43329c6d403d27a361b8

File: scapy/fields.py
Comment (gpotter2) [reply scapy/fields.py:257]: Here's what mypy returns:

```
mypy run-test: commands[0] | python .config/mypy/mypy_check.py
scapy\fields.py:2541: error: Argument 1 of "__eq__" is incompatible with supertype "object"; supertype defines the argument type as "object"
scapy\fields.py:2541: note: This violates the Liskov substitution principle
scapy\fields.py:2541: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
scapy\fields.py:2541: note: It is recommended for "__eq__" to work with arbitrary objects, for example:
scapy\fields.py:2541: note:     def __eq__(self, other: object) -> bool:
scapy\fields.py:2541: note:         if not isinstance(other, FlagValue):
scapy\fields.py:2541: note:             return NotImplemented
scapy\fields.py:2541: note:         return <logic to compare two FlagValue instances>
```

**Comment by gpotter2** [comment]
I'm pretty sure I covered all your comments @guedou.
@guedou @p-l- I think it might be nice to start releasing 2.4.4 before merging this (?).

**Comment by p-l-** [comment]
@gpotter2 agreed!

File: scapy/fields.py
Comment (guedou) [reply scapy/fields.py:257]: I did some experiments and it seems that there is a `--strict-equality` mypy flag that could help detect things like `Field(...) == 3`. According to mypy documentation and the error message above, we should type ours `__eq__()` and `__ne__()` like:
```
def __eq__(self, other: object) -> bool:
    if not isinstance(other, FlagValue):
       return NotImplemented
    return <logic to compare two FlagValue instances>
```https://stackoverflow.com/questions/37557411/why-does-defining-the-argument-types-for-eq-throw-a-mypy-type-error

I will understand that you don't want to add the `if` block, but could you change `Any` to `object`?

**Comment by guedou** [comment]
I tried the `mypy` `--strict-equality` option. It is able to catch things like `"scapy" == 2807`. but I did not manage to catch `IP() == 42` yet.

File: scapy/fields.py
Comment (gpotter2) [inline scapy/fields.py:3231]: I had quite some trouble making BitScalingField work in the first place on both Python 2 and Python 3 so its typing will be mostly ignored.

**Comment by gpotter2** [comment]
@guedou @p-l- Hi, this is ready for review. I would like to have this merged as soon as possible because it's a pain to keep up to date with master. Thanks

File: scapy/fields.py
Comment (guedou) [reply scapy/fields.py:257]: @gpotter2 did you try this?

Comment (gpotter2) [reply scapy/fields.py:259]: Except we don't want to do that, since we want to have
```
>>> FlagValue(1, ["a", "b", "c"])
<Flag 1 (a)>
>>> FlagValue(1, ["a", "b", "c"]) == "a"
True
>>> FlagValue(1, ["a", "b", "c"]) == 1
True
```

Comment (guedou) [reply scapy/fields.py:259]: I really believe that we want to enforce strict type checking. Otherwise, this makes typing quite unhelpful considering the work involved in annotating Scapy. The FlagValue is a valid counter example, but we should handle it separately.

I am OK with merging this PR, as soon as we move this stricter type checking discussions to a dedicated issue =)

Comment (gpotter2) [reply scapy/fields.py:259]: Sounds good, let's do that in a future PR. See https://github.com/secdev/scapy/issues/2823

**Comment by guedou** [comment]
@p-l- what's your call? I am ok with merging this PR as it is.

**Comment by gpotter2** [comment]
Merge ? Ideally we could get hacktoberfest people to work on typing Scapy... 😄

**Comment by gpotter2** [comment]
@p-l- Could you please review this? Thanks

**Comment by gpotter2** [comment]
Merge?

**Review by p-l- — APPROVED** [review]

**Pull request merged**
